### ISSUE TITLE
Inversion en cas de non linéarité

### ISSUE THREAD

**Issue by laem** [betagouv/syso#443]
Les cotisations des indépendants, comme celles du régime général, ne sont pas linéaires ! 

Pour les indépendants, c'est la retraite complémentaire interprofessionnelle des libéraux qui foire, avec un saut de cotisation pour chaque tranche très conséquent (5000 € !!), pour les salariés c'est les réductions alloc familiales et surtout la nouvelle réduction maladie de 6% pour les salaires > 2.5 SMIC. Cette dernière est très visible.

Il faut trouver un moyen d'expliquer à l'utilisateur :art: que **certaines saisies sont impossibles :stop_sign: **.

> Cette montant, 4546€, n'est pas possible : le calcul des cotisations sociales comprendre des "sauts" (pour 1€ brut de plus, le salaire total augmente de 300€). Essayez de saisir un montant inférieur à 4100€ ou supérieur à 4600€.
.

**Comment by mquandalle** [comment]
Le régime complémentaire de la Cipav permet de cotiser à la tranche immédiatement au dessus de son revenu sur demande. On a donc le choix entre une tranche à `x` € ou une tranche `x + 5000` € (5000 représentant le cas extrême, la plupart des sauts de tranches sont plus faibles).

Je n'ai pas fait les calculs mais dans la mesure ou l'on peut "déplacer" le saut de tranche, il me semble que cela doit parfois résoudre les inversions net vers brut.

Cela semble assez difficile à implémenter cependant, et potentiellement source de confusion. Pas certain que cela soit préférable par rapport à un message type "ce montant est impossible".

**Comment by mquandalle** [comment]
Je suis tombé au hasard sur un calcul noté comme impossible. Sur le simulateur des cotisations de l'indépendant (https://mon-entreprise.fr/sécurité-sociale/indépendant), rentrer un CA de 80000€ donne l'erreur suivante :

![erreur calcul impossible](https://user-images.githubusercontent.com/1730702/58317736-b73b2e00-7e16-11e9-810d-1461fbccb063.png)

Le montant n'est pas trop faible, donc c'est que le calcul est impossible (peut-être faudrait-il avoir un message d'erreur spécifique pour chacune de ces deux erreurs). Je ne sais pas si c'est lié à cette issue ou si c'est un bug, mais je trouve assez étrange qu'on ne puisse pas calculer les cotisations à partir d'un CA donné.

**Comment by laem** [comment]
Oui, c'est du au fait qu'il y a un barème de retraite complémentaire avec de gros sauts, et une inversion. Dans la vraie vie, le dirigeant choisit un revenu pro qui lui convient, par exemple (chiffres fictifs) 50k, calcule les cotisations, ce qui donne un revenu total de 75k, puis (s'il est à l'IS) garde les 5k restants dans la caisse. S'il choisit de se verser 50001k, le barème de retraite complémentaire fera passer son revenu total à 82k :laughing: Je trouve ça gros comme décision des legislateurs, donc je suis ouvert à une remise en question...

**Comment by mquandalle** [comment]
> Dans la vraie vie, le dirigeant choisit un revenu pro qui lui convient [puis garde le reste]

C'est une approche possible, mais même à l'IS, il peut aussi arriver que le dirigeant veuille sortir le maximum d'argent possible de sa structure (investissement immobilier à titre personnel, anticipation d'augmentation de fiscalité, anticipation de liquidation de la société, etc.). Dans ce cas le raisonnement est : « j'ai 80 000 à distribuer comment répartir entre salaire et cotisations (et dividende le cas échéant) ». 

Alors après effectivement, le barème de retraite complémentaire rend la réponse à cette question particulièrement délicate, et le plus simple semble effectivement d'utiliser les dividendes comme variables d'ajustement pour éviter les sauts de tranches.

Une autre possibilité est de se verser moins de salaire mais de sur-cotiser à la tranche d'au-dessus sur la retraite complémentaire (avec plus de points retraites à la clé et donc plus de salaire différé), avec cette solution il me semble que l'on arrive à distribuer exactement 80000€ par exemple, mais prendre ce type de décision (cotisation à la tranche supérieure) dans la calculateur sans la proposer à l'utilisateur et uniquement pour trouver une solution à un calcul n'aurait pas beaucoup de sens...

Bon de toute façon, il y a une réforme des retraites qui arrive bientôt, j'imagine que ce système de tranches va disparaître.

**Comment by laem** [comment]
Exactement ! 
(La réforme des retraites, faut voir quand elle va attérir par contre...)

**Comment by mquandalle** [comment]
On a clarifié les messages d'erreurs quand cette situation se présente, je ferme car c'est un bug de la législation pas de notre code
